Using stop, wait and retry custom callbacks with Retrying causes mypy failures

I am using custom callbacks for `stop` and `wait` in the form of functions accepting a retry state, passing them to the `Retrying` context manager, as documented in [the Custom Callbacks section](https://tenacity.readthedocs.io/en/latest/#other-custom-callbacks). With the latest version of tenacity, this is now causing `mypy` failures since `BaseRetrying` [now only accepts `wait` and `stop` callbacks](https://github.com/jd/tenacity/blob/a7f548520e4ee871ad8aeb354ecfa1a324c8ca19/tenacity/__init__.py#L228-L230) that are function objects of a class inheriting from `wait_base` and `stop_base`. The same will also happen if a `retry` custom callback was provided.

These changes were introduced in #304. 

Since wait_base and stop_base aren't exposed as part of the public API, I believe this should be updated in order to enable using `Retrying` directly with `stop`, `wait` and `retry` custom callbacks as documented.

Small example causing `mypy` failures:
``` python
import tenacity

def my_stop(retry_state: tenacity.RetryCallState) -> bool:
    return True


def my_retry(retry_state: tenacity.RetryCallState) -> bool:
    return False


def my_wait(retry_state: tenacity.RetryCallState) -> bool:
    return False


for attempt in tenacity.Retrying(retry=my_retry, wait=my_wait, stop=my_stop):
    with attempt:
        pass
```

`mypy` output:

``` shell
$ python3.9 -m mypy test.py 
test.py:15: error: Argument "retry" to "Retrying" has incompatible type "Callable[[RetryCallState], bool]"; expected "retry_base"
test.py:15: error: Argument "wait" to "Retrying" has incompatible type "Callable[[RetryCallState], bool]"; expected "wait_base"
test.py:15: error: Argument "stop" to "Retrying" has incompatible type "Callable[[RetryCallState], bool]"; expected "stop_base"
Found 3 errors in 1 file (checked 1 source file)
```

If we change the same example to use the `@retry` decorator instead of the `Retrying` context manager, `mypy` reports no errors:

``` python
import tenacity

def my_stop(retry_state: tenacity.RetryCallState) -> bool:
    return True


def my_retry(retry_state: tenacity.RetryCallState) -> bool:
    return False


def my_wait(retry_state: tenacity.RetryCallState) -> bool:
    return False

@tenacity.retry(wait=my_wait, stop=my_stop, retry=my_retry)
def retryable_function():
    pass

retryable_function()
```

`mypy` output:
```
$ python3.9 -m mypy test.py 
Success: no issues found in 1 source file
```

--- a/tenacity/__init__.py
+++ b/tenacity/__init__.py
@@ -86,8 +86,9 @@ except ImportError:
 if t.TYPE_CHECKING:
     import types
 
-    from .wait import wait_base
-    from .stop import stop_base
+    from .retry import RetryBaseT
+    from .stop import StopBaseT
+    from .wait import WaitBaseT
 
 
 WrappedFn = t.TypeVar("WrappedFn", bound=t.Callable[..., t.Any])
@@ -217,9 +218,9 @@ class BaseRetrying(ABC):
     def __init__(
         self,
         sleep: t.Callable[[t.Union[int, float]], None] = sleep,
-        stop: "stop_base" = stop_never,
-        wait: "wait_base" = wait_none(),
-        retry: retry_base = retry_if_exception_type(),
+        stop: "StopBaseT" = stop_never,
+        wait: "WaitBaseT" = wait_none(),
+        retry: "RetryBaseT" = retry_if_exception_type(),
         before: t.Callable[["RetryCallState"], None] = before_nothing,
         after: t.Callable[["RetryCallState"], None] = after_nothing,
         before_sleep: t.Optional[t.Callable[["RetryCallState"], None]] = None,
@@ -242,8 +243,8 @@ class BaseRetrying(ABC):
     def copy(
         self,
         sleep: t.Union[t.Callable[[t.Union[int, float]], None], object] = _unset,
-        stop: t.Union["stop_base", object] = _unset,
-        wait: t.Union["wait_base", object] = _unset,
+        stop: t.Union["StopBaseT", object] = _unset,
+        wait: t.Union["WaitBaseT", object] = _unset,
         retry: t.Union[retry_base, object] = _unset,
         before: t.Union[t.Callable[["RetryCallState"], None], object] = _unset,
         after: t.Union[t.Callable[["RetryCallState"], None], object] = _unset,
@@ -337,14 +338,14 @@ class BaseRetrying(ABC):
             return DoAttempt()
 
         is_explicit_retry = fut.failed and isinstance(fut.exception(), TryAgain)
-        if not (is_explicit_retry or self.retry(retry_state=retry_state)):
+        if not (is_explicit_retry or self.retry(retry_state)):
             return fut.result()
 
         if self.after is not None:
             self.after(retry_state)
 
         self.statistics["delay_since_first_attempt"] = retry_state.seconds_since_start
-        if self.stop(retry_state=retry_state):
+        if self.stop(retry_state):
             if self.retry_error_callback:
                 return self.retry_error_callback(retry_state)
             retry_exc = self.retry_error_cls(fut)
@@ -353,7 +354,7 @@ class BaseRetrying(ABC):
             raise retry_exc from fut.exception()
 
         if self.wait:
-            sleep = self.wait(retry_state=retry_state)
+            sleep = self.wait(retry_state)
         else:
             sleep = 0.0
         retry_state.next_action = RetryAction(sleep)

--- a/tenacity/retry.py
+++ b/tenacity/retry.py
@@ -36,6 +36,9 @@ class retry_base(abc.ABC):
         return retry_any(self, other)
 
 
+RetryBaseT = typing.Union[retry_base, typing.Callable[["RetryCallState"], bool]]
+
+
 class _retry_never(retry_base):
     """Retry strategy that never rejects any result."""
 

--- a/tenacity/stop.py
+++ b/tenacity/stop.py
@@ -38,6 +38,9 @@ class stop_base(abc.ABC):
         return stop_any(self, other)
 
 
+StopBaseT = typing.Union[stop_base, typing.Callable[["RetryCallState"], bool]]
+
+
 class stop_any(stop_base):
     """Stop if any of the stop condition is valid."""
 

--- a/tenacity/wait.py
+++ b/tenacity/wait.py
@@ -41,6 +41,9 @@ class wait_base(abc.ABC):
         return self.__add__(other)
 
 
+WaitBaseT = typing.Union[wait_base, typing.Callable[["RetryCallState"], bool]]
+
+
 class wait_fixed(wait_base):
     """Wait strategy that waits a fixed amount of time between each retry."""
 

